src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -154,13 +261,11 @@ def infer_metadata(
                 if len(quoted_identifiers) != len(set(quoted_identifiers)):
                     quoted_identifiers = None
         # If source_plan is a SelectableEntity or SelectStatement, SQL simplifier is enabled
-        if isinstance(source_plan, SelectableEntity):
+        elif isinstance(source_plan, SelectableEntity):
             if source_plan.attributes is not None:
                 attributes = source_plan.attributes
         elif isinstance(source_plan, SelectStatement):
-            # When attributes is cached on source_plan, just use it
-            if source_plan.attributes is not None:
-                attributes = source_plan.attributes
+            attributes = _extract_selectable_attributes(source_plan)
```

Comment:
I see that this function is recursive so it can invoke other selectable nodes. Can we re-use the same function in any of the if-else branches above?